adam-core/src/main/scala/org/bdgenomics/adam/rdd/ShuffleRegionJoin.scala

```diff
@@ -263,6 +263,30 @@ private[rdd] case class ManualRegionPartitioner(partitions: Int) extends Partiti
   }
 }
 
+private class AppendableIterator[T] extends Iterator[T] {
+  var iterators: ListBuffer[Iterator[T]] = ListBuffer.empty
```

Comment:
Do you need to worry about thread safety?  Would immutable copy-on-write help here?